diff --git a/color-schemes/render/ash.json b/color-schemes/render/ash.json
new file mode 100644
index 000000000..eb28280a3
--- /dev/null
+++ b/color-schemes/render/ash.json
@@ -0,0 +1,21 @@
+{
+    "name" : "Ash",
+    "index" : 2000,
+    "show-in-gui" : true,
+    "description" : "monochrome version of Cornfield",
+
+    "colors" : {
+        "background" :         "#fdfdfd",
+        "highlight" :          "#51ff5180",
+        "axes-color" :         "#000000",
+        "opencsg-face-front" : "#d1d1d1",
+        "opencsg-face-back" :  "#b8b8b8",
+        "cgal-face-front" :    "#d1d1d1",
+        "cgal-face-back" :     "#b8b8b8",
+        "cgal-face-2d" :       "#939393",
+        "cgal-edge-front" :    "#e5e5e5",
+        "cgal-edge-back" :     "#c5c5c5",
+        "cgal-edge-2d" :       "#363636",
+        "crosshair" :          "#1b1b1b"
+    }
+}
diff --git a/color-schemes/render/beforedawn.json b/color-schemes/render/beforedawn.json
index d4d98bba9..0f05c2f85 100644
--- a/color-schemes/render/beforedawn.json
+++ b/color-schemes/render/beforedawn.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#333333",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#c1c1c1",
         "opencsg-face-front" : "#cccccc",
         "opencsg-face-back" :  "#5563dd",
diff --git a/color-schemes/render/deepocean.json b/color-schemes/render/deepocean.json
index 9ce164fdc..f5179c583 100644
--- a/color-schemes/render/deepocean.json
+++ b/color-schemes/render/deepocean.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#333333",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#c1c1c1",
         "opencsg-face-front" : "#eeeeee",
         "opencsg-face-back" :  "#0babc8",
diff --git a/color-schemes/render/gencolorscheme.py b/color-schemes/render/gencolorscheme.py
new file mode 100755
index 000000000..de90e6e3e
--- /dev/null
+++ b/color-schemes/render/gencolorscheme.py
@@ -0,0 +1,55 @@
+#!/usr/bin/python3
+# print openscad color scheme
+
+import colorsys
+
+# print rgb as hex string
+# r=0..255, g=0..255, b=0.255
+def rgb(r, g, b):
+  hexval = hex(256*256*256+r*256*256+g*256+b)
+  return hexval[3:9]
+
+# print rgb as hex string
+# h=0..360, s=0..100, v=0..100
+def hsv(h, s, v):
+  color = tuple(int(round(i * 255)) for i in colorsys.hsv_to_rgb(h/360.0,s/100.0,v/100.0))
+  return rgb(color[0], color[1], color[2])
+
+# color scheme
+background =         rgb(255,255,255)
+highlight =          hsv(60,50,80)
+axes_color =         hsv(60,50,50)
+opencsg_face_front = hsv(75,50,80)
+opencsg_face_back =  hsv(285,50,80)
+cgal_face_front =    hsv(75,50,90)
+cgal_face_back =     hsv(285,50,80)
+cgal_face_2d =       hsv(90,50,90)
+cgal_edge_front =    hsv(75,50,90)
+cgal_edge_back =     hsv(290,50,80)
+cgal_edge_2d =       hsv(275,50,80)
+crosshair =          hsv(0,0,50)
+
+print (f'''{{
+    "name" : "Red Cyan Glasses",
+    "index" : 2001,
+    "show-in-gui" : true,
+    "description" : "3d anaglyph",
+    "_comment" : "created by gencolorscheme.py",
+
+    "colors" : {{
+        "background" :         "#{background}",
+        "highlight" :          "#{highlight}80",
+        "axes-color" :         "#{axes_color}",
+        "opencsg-face-front" : "#{opencsg_face_front}",
+        "opencsg-face-back" :  "#{opencsg_face_back}",
+        "cgal-face-front" :    "#{cgal_face_front}",
+        "cgal-face-back" :     "#{cgal_face_back}",
+        "cgal-face-2d" :       "#{cgal_face_2d}",
+        "cgal-edge-front" :    "#{cgal_edge_front}",
+        "cgal-edge-back" :     "#{cgal_edge_back}",
+        "cgal-edge-2d" :       "#{cgal_edge_2d}",
+        "crosshair" :          "#{crosshair}"
+    }}
+}}''')
+
+#not truncated
diff --git a/color-schemes/render/metallic.json b/color-schemes/render/metallic.json
index 838558769..08e91129e 100644
--- a/color-schemes/render/metallic.json
+++ b/color-schemes/render/metallic.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#aaaaff",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#222233",
         "opencsg-face-front" : "#ddddff",
         "opencsg-face-back" :  "#dd22dd",
diff --git a/color-schemes/render/monotone.json b/color-schemes/render/monotone.json
index df1a79aa7..06cb8f8d9 100644
--- a/color-schemes/render/monotone.json
+++ b/color-schemes/render/monotone.json
@@ -6,6 +6,7 @@
 
     "colors" : {
         "background" :         "#ffffe5",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#191916",
         "opencsg-face-front" : "#f9d72c",
         "opencsg-face-back" :  "#f9d72c",
diff --git a/color-schemes/render/nature.json b/color-schemes/render/nature.json
index 6250e7a33..b71ec0b44 100644
--- a/color-schemes/render/nature.json
+++ b/color-schemes/render/nature.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#fafafa",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#323232",
         "opencsg-face-front" : "#16a085",
         "opencsg-face-back" :  "#dbf4da",
diff --git a/color-schemes/render/redcyanglasses.json b/color-schemes/render/redcyanglasses.json
new file mode 100644
index 000000000..2087ed474
--- /dev/null
+++ b/color-schemes/render/redcyanglasses.json
@@ -0,0 +1,22 @@
+{
+    "name" : "Red Cyan Glasses",
+    "index" : 2001,
+    "show-in-gui" : true,
+    "description" : "3d anaglyph",
+    "_comment" : "created by gencolorscheme.py",
+
+    "colors" : {
+        "background" :         "#ffffff",
+        "highlight" :          "#cccc6680",
+        "axes-color" :         "#808040",
+        "opencsg-face-front" : "#b3cc66",
+        "opencsg-face-back" :  "#b366cc",
+        "cgal-face-front" :    "#c9e673",
+        "cgal-face-back" :     "#b366cc",
+        "cgal-face-2d" :       "#ace673",
+        "cgal-edge-front" :    "#c9e673",
+        "cgal-edge-back" :     "#bb66cc",
+        "cgal-edge-2d" :       "#a266cc",
+        "crosshair" :          "#808080"
+    }
+}
diff --git a/color-schemes/render/solarized.json b/color-schemes/render/solarized.json
index 0fd5002b1..f7c482d4b 100644
--- a/color-schemes/render/solarized.json
+++ b/color-schemes/render/solarized.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#fdf6e3",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#191816",
         "opencsg-face-front" : "#b58800",
         "opencsg-face-back" :  "#882233",
diff --git a/color-schemes/render/starnight.json b/color-schemes/render/starnight.json
index 88a9510c3..6e076a377 100644
--- a/color-schemes/render/starnight.json
+++ b/color-schemes/render/starnight.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#000000",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#e5e5e5",
         "opencsg-face-front" : "#ffffe0",
         "opencsg-face-back" :  "#00ffff",
diff --git a/color-schemes/render/sunset.json b/color-schemes/render/sunset.json
index bd7ae26b1..c22880d37 100644
--- a/color-schemes/render/sunset.json
+++ b/color-schemes/render/sunset.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#aa4444",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#220d0d",
         "opencsg-face-front" : "#ffaaaa",
         "opencsg-face-back" :  "#882233",
diff --git a/color-schemes/render/tomorrow-night.json b/color-schemes/render/tomorrow-night.json
index 8b7966c0a..90363a2b9 100644
--- a/color-schemes/render/tomorrow-night.json
+++ b/color-schemes/render/tomorrow-night.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#1d1f21",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#e8e8e8",
         "opencsg-face-front" : "#81a2be",
         "opencsg-face-back" :  "#de935f",
diff --git a/color-schemes/render/tomorrow.json b/color-schemes/render/tomorrow.json
index 559e23fe3..1b098cbe1 100644
--- a/color-schemes/render/tomorrow.json
+++ b/color-schemes/render/tomorrow.json
@@ -5,6 +5,7 @@
 
     "colors" : {
         "background" :         "#f8f8f8",
+        "highlight" :          "#ff515180",
         "axes-color" :         "#181818",
         "opencsg-face-front" : "#4271ae",
         "opencsg-face-back" :  "#f5871f",
diff --git a/images/Anaglyph-32.png b/images/Anaglyph-32.png
new file mode 100644
index 000000000..b7bc5c4a0
Binary files /dev/null and b/images/Anaglyph-32.png differ
diff --git a/openscad.qrc b/openscad.qrc
index 0b143032c..247305a3b 100644
--- a/openscad.qrc
+++ b/openscad.qrc
@@ -106,6 +106,7 @@
     <file>images/SVG-white.png</file>
     <file>images/PDF.png</file>
     <file>images/PDF-white.png</file>
+    <file>images/Anaglyph-32.png</file>
     <file>sounds/complete.wav</file>
   </qresource>
 </RCC>
diff --git a/src/Camera.cc b/src/Camera.cc
index 4c13820d6..baf4d3ae0 100644
--- a/src/Camera.cc
+++ b/src/Camera.cc
@@ -1,22 +1,29 @@
 #include "Camera.h"
+#include "QSettingsCached.h"
 #include "rendersettings.h"
 #include "printutils.h"
 #include "degree_trig.h"
 
 static const double DEFAULT_DISTANCE = 140.0;
 static const double DEFAULT_FOV = 22.5;
+static const float DEFAULT_EYE_SEPARATION = 65.0; // 65 mm avg. between human eyes
 
 Camera::Camera() :
 	projection(ProjectionType::PERSPECTIVE), fov(DEFAULT_FOV), viewall(false), autocenter(false)
 {
 	PRINTD("Camera()");
 
+	QSettingsCached settings;
 	// gimbal cam values
 	resetView();
 
 	pixel_width = RenderSettings::inst()->img_width;
 	pixel_height = RenderSettings::inst()->img_height;
 	locked = false;
+	anaglyph = false;
+	eyeSeparation = settings.value("3dview/eyeSeparation", DEFAULT_EYE_SEPARATION).toDouble();
+	if ((eyeSeparation <= 0.0) || (eyeSeparation > 100.0)) eyeSeparation = DEFAULT_EYE_SEPARATION;
+	std::cerr << "eye separation: " << this->eyeSeparation << std::endl;
 }
 
 void Camera::setup(std::vector<double> params)
@@ -65,6 +72,16 @@ void Camera::viewAll(const BoundingBox &bbox)
 	}
 }
 
+void Camera::zoomAnaglyph(int zoom)
+{
+	QSettingsCached settings;
+	this->eyeSeparation *= pow(0.95, -zoom / 120.0);
+	if (eyeSeparation >= 100.0) eyeSeparation = 100.0;
+	else if (eyeSeparation <= 0.0) eyeSeparation = DEFAULT_EYE_SEPARATION;
+	settings.setValue("3dview/eyeSeparation", QString::fromStdString(std::to_string(this->eyeSeparation)));
+	std::cerr << "eye separation: " << this->eyeSeparation << std::endl;
+}
+
 void Camera::zoom(int zoom, bool relative)
 {
     if (relative) {
diff --git a/src/Camera.h b/src/Camera.h
index 4cd2859c3..b770f6d24 100644
--- a/src/Camera.h
+++ b/src/Camera.h
@@ -29,6 +29,7 @@ public:
 	void setup(std::vector<double> params);
 	void gimbalDefaultTranslate();
 	void setProjection(ProjectionType type);
+	void zoomAnaglyph(int delta);
 	void zoom(int delta, bool relative);
 	double zoomValue() const;
 	double fovValue() const;
@@ -67,6 +68,10 @@ public:
 	// (--camera option in commandline mode)
 	bool locked;
 
+	// true if in anaglyph 3D stereo view
+	bool anaglyph;
+	double eyeSeparation;
+
 protected:
 	// Perspective settings
 	double viewer_distance;
diff --git a/src/GLView.cc b/src/GLView.cc
index 766387736..43631fffb 100644
--- a/src/GLView.cc
+++ b/src/GLView.cc
@@ -115,8 +115,104 @@ void GLView::setupCamera() const
   glRotated(cam.object_rot.z(), 0.0, 0.0, 1.0);
 }
 
+/* set up camera for left or right eye */
+/* after https://quiescentspark.blogspot.com/2011/05/rendering-3d-anaglyph-in-opengl.html */
+void GLView::setup3dCamera(bool leftCamera) {
+  float left, right, bottom, top;
+  float convergence = (float)cam.zoomValue();
+  float aspectRatio = aspectratio;
+  float fov = cam.fov;
+  float nearClippingDistance = 0.1 * cam.zoomValue();
+  float farClippingDistance = 100.0 * cam.zoomValue();
+
+  // calculate frustum
+  top = nearClippingDistance * tan_degrees(fov / 2.0);
+  bottom = -top;
+  float a = aspectRatio * tan_degrees(fov / 2.0) * convergence;
+  float b = a - cam.eyeSeparation / 2;
+  float c = a + cam.eyeSeparation / 2;
+  if (leftCamera) { // left camera
+    left = -b * nearClippingDistance / convergence;
+    right = c * nearClippingDistance / convergence;
+  }
+  else { // right camera
+    left = -c * nearClippingDistance / convergence;
+    right = b * nearClippingDistance / convergence;
+  }
+
+  glMatrixMode(GL_PROJECTION);
+  glLoadIdentity();
+  glFrustum(left, right, bottom, top, nearClippingDistance, farClippingDistance);
+  glMatrixMode(GL_MODELVIEW);
+  glLoadIdentity();
+  if (leftCamera) { // left camera
+    glTranslatef(cam.eyeSeparation / 2, 0.0, 0.0);
+  }
+  else { // right camera
+    glTranslatef(-cam.eyeSeparation / 2, 0.0, 0.0);
+  }
+  gluLookAt(0.0, -cam.zoomValue(), 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0);
+  glRotated(cam.object_rot.x(), 1.0, 0.0, 0.0);
+  glRotated(cam.object_rot.y(), 0.0, 1.0, 0.0);
+  glRotated(cam.object_rot.z(), 0.0, 0.0, 1.0);
+}
+
+void GLView::paintOnce()
+{
+  auto crosshaircol = ColorMap::getColor(*this->colorscheme, RenderColor::CROSSHAIR_COLOR);
+  // The crosshair should be fixed at the center of the viewport...
+  if (showcrosshairs) GLView::showCrosshairs(crosshaircol);
+  glTranslated(cam.object_trans.x(), cam.object_trans.y(), cam.object_trans.z());
+  // ...the axis lines need to follow the object translation.
+  auto axescolor = ColorMap::getColor(*this->colorscheme, RenderColor::AXES_COLOR);
+  if (showaxes) GLView::showAxes(axescolor);
+  // mark the scale along the axis lines
+  if (showaxes && showscale) GLView::showScalemarkers(axescolor);
+
+  glEnable(GL_LIGHTING);
+  glDepthFunc(GL_LESS);
+  glCullFace(GL_BACK);
+  glDisable(GL_CULL_FACE);
+  glLineWidth(2);
+  glColor3d(1.0, 0.0, 0.0);
+
+  if (this->renderer) {
+#if defined(ENABLE_OPENCSG)
+    // FIXME: This belongs in the OpenCSG renderer, but it doesn't know about this ID yet
+    OpenCSG::setContext(this->opencsg_id);
+#endif
+    this->renderer->prepare(showfaces, showedges);
+    this->renderer->draw(showfaces, showedges);
+  }
+
+  glDisable(GL_LIGHTING);
+}
+
+void GLView::anaglyphHalfColor() {
+  glDisable(GL_LIGHTING);
+  auto bgcol = ColorMap::getColor(*this->colorscheme, RenderColor::BACKGROUND_COLOR);
+  glClearColor(bgcol[0], bgcol[1], bgcol[2], 1.0);
+
+  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+  setup3dCamera(true);
+  glColorMask(true, false, false, false); // red
+  paintOnce();
+
+  glClear(GL_DEPTH_BUFFER_BIT);
+  setup3dCamera(false);
+  glColorMask(false, true, true, false); // cyan
+  paintOnce();
+
+  glColorMask(true, true, true, true);
+}
+
 void GLView::paintGL()
 {
+  if (cam.anaglyph) {
+    anaglyphHalfColor();
+    return;
+  }
+
   glDisable(GL_LIGHTING);
   auto bgcol = ColorMap::getColor(*this->colorscheme, RenderColor::BACKGROUND_COLOR);
   auto axescolor = ColorMap::getColor(*this->colorscheme, RenderColor::AXES_COLOR);
diff --git a/src/GLView.h b/src/GLView.h
index 1fcd31bc8..2e6f49b13 100644
--- a/src/GLView.h
+++ b/src/GLView.h
@@ -2,20 +2,20 @@
 
 /* GLView: A basic OpenGL rectangle for rendering images.
 
-This class is inherited by:
+	 This class is inherited by:
 
-*QGLview - for Qt GUI
-*OffscreenView - for offscreen rendering, in tests and from command-line
-(This class is also overridden by NULLGL.cc for special experiments)
+ * QGLview - for Qt GUI
+ * OffscreenView - for offscreen rendering, in tests and from command-line
+	 (This class is also overridden by NULLGL.cc for special experiments)
 
-The view assumes either a Gimbal Camera (rotation,translation,distance)
-or Vector Camera (eye,center/target) is being used. See Camera.h. The
-cameras are not kept in sync.
+	 The view assumes either a Gimbal Camera (rotation,translation,distance)
+	 or Vector Camera (eye,center/target) is being used. See Camera.h. The
+	 cameras are not kept in sync.
 
-QGLView only uses GimbalCamera while OffscreenView can use either one.
-Some actions (showCrossHairs) only work properly on Gimbal Camera.
+	 QGLView only uses GimbalCamera while OffscreenView can use either one.
+	 Some actions (showCrossHairs) only work properly on Gimbal Camera.
 
-*/
+ */
 
 #include <Eigen/Core>
 #include <Eigen/Geometry>
@@ -29,7 +29,7 @@ class GLView
 {
 public:
 	GLView();
-	void setRenderer(class Renderer* r);
+	void setRenderer(class Renderer *r);
 	Renderer *getRenderer() const { return this->renderer; }
 
 	void initializeGL();
@@ -39,6 +39,10 @@ public:
 	void setCamera(const Camera &cam);
 	void setupCamera() const;
 
+	void setup3dCamera(bool leftCamera);
+	void paintOnce();
+	void anaglyphHalfColor();
+
 	void setColorScheme(const ColorScheme &cs);
 	void setColorScheme(const std::string &cs);
 	void updateColorScheme();
@@ -58,7 +62,7 @@ public:
 	virtual std::string getRendererInfo() const = 0;
 	virtual float getDPI() { return 1.0f; }
 
-	virtual ~GLView(){};
+	virtual ~GLView() {}
 
 	Renderer *renderer;
 	const ColorScheme *colorscheme;
diff --git a/src/MainWindow.h b/src/MainWindow.h
index aaed0e139..25f08b381 100644
--- a/src/MainWindow.h
+++ b/src/MainWindow.h
@@ -299,6 +299,7 @@ public slots:
 	void viewCenter();
 	void viewPerspective();
 	void viewOrthogonal();
+	void viewAnaglyph();
 	void viewTogglePerspective();
 	void viewResetView();
 	void viewAll();
diff --git a/src/MainWindow.ui b/src/MainWindow.ui
index c3ea8e0e9..93eede35f 100644
--- a/src/MainWindow.ui
+++ b/src/MainWindow.ui
@@ -251,6 +251,7 @@
          <addaction name="separator"/>
          <addaction name="viewActionPerspective"/>
          <addaction name="viewActionOrthogonal"/>
+         <addaction name="viewActionAnaglyph"/>
          <addaction name="separator"/>
          <addaction name="viewActionShowAxes"/>
          <addaction name="viewActionShowScaleProportional"/>
@@ -413,6 +414,7 @@
     <addaction name="separator"/>
     <addaction name="viewActionPerspective"/>
     <addaction name="viewActionOrthogonal"/>
+    <addaction name="viewActionAnaglyph"/>
     <addaction name="separator"/>
     <addaction name="viewActionHideEditorToolBar"/>
     <addaction name="viewActionHide3DViewToolBar"/>
@@ -1302,6 +1304,24 @@
     <string>&amp;Orthogonal</string>
    </property>
   </action>
+  <action name="viewActionAnaglyph">
+   <property name="checkable">
+    <bool>true</bool>
+   </property>
+   <property name="icon">
+    <iconset resource="../openscad.qrc">
+     <normaloff>:/images/Anaglyph-32.png</normaloff>:/images/Anaglyph-32.png</iconset>
+   </property>
+   <property name="text">
+    <string>&amp;Stereo Vision (Anaglyph)</string>
+   </property>
+   <property name="toolTip">
+    <string>Stereo Vision (Anaglyph) - requires special red-cyan glasses</string>
+   </property>
+   <property name="shortcut">
+    <string>Alt+S</string>
+   </property>
+  </action>
   <action name="windowActionHideConsole">
    <property name="checkable">
     <bool>true</bool>
diff --git a/src/QGLView.cc b/src/QGLView.cc
index bce613073..c2d7f0b6c 100644
--- a/src/QGLView.cc
+++ b/src/QGLView.cc
@@ -324,7 +324,10 @@ void QGLView::wheelEvent(QWheelEvent *event)
 {
 	const auto pos = event->pos();
 	const int v = event->angleDelta().y();
-	if (this->mouseCentricZoom) {
+	if (QGuiApplication::keyboardModifiers().testFlag(Qt::ControlModifier)) {
+		cam.zoomAnaglyph(v);
+		updateGL();
+	} else if (this->mouseCentricZoom) {
 		zoomCursor(pos.x(), pos.y(), v);
 	} else {
 		zoom(v, true);
diff --git a/src/QGLView.h b/src/QGLView.h
index 195852972..33325b762 100644
--- a/src/QGLView.h
+++ b/src/QGLView.h
@@ -41,6 +41,7 @@ public:
 	void setOrthoMode(bool enabled);
 	bool showScaleProportional() const { return this->showscale; }
 	void setShowScaleProportional(bool enabled) { this->showscale = enabled; }
+	void setAnaglyphMode(bool enabled) { this->cam.anaglyph = enabled; }
 	std::string getRendererInfo() const override;
 	float getDPI() override { return this->devicePixelRatio(); }
 
diff --git a/src/builtin.cc b/src/builtin.cc
index aa857908a..6f20d6b54 100644
--- a/src/builtin.cc
+++ b/src/builtin.cc
@@ -113,6 +113,7 @@ Builtins::Builtins()
 	this->assignments.emplace_back(new Assignment("$vpr", make_shared<Literal>(VectorType(0.0, 0.0, 0.0))) );
 	this->assignments.emplace_back(new Assignment("$vpd", make_shared<Literal>(500)) );
 	this->assignments.emplace_back(new Assignment("$vpf", make_shared<Literal>(22.5)) );
+	this->assignments.emplace_back(new Assignment("$anaglyph", make_shared<Literal>(Value::undefined.clone())) ); // true if rendering anaglyph 3d
 }
 
 void Builtins::initKeywordList()
diff --git a/src/colormap.cc b/src/colormap.cc
index 6b794be17..71b035635 100644
--- a/src/colormap.cc
+++ b/src/colormap.cc
@@ -37,6 +37,7 @@ RenderColorScheme::RenderColorScheme() : _path("")
   _show_in_gui = true;
 
   _color_scheme.insert(ColorScheme::value_type(RenderColor::BACKGROUND_COLOR, Color4f(0xff, 0xff, 0xe5)));
+  _color_scheme.insert(ColorScheme::value_type(RenderColor::HIGHLIGHT_COLOR, Color4f(0xff, 0x51, 0x51, 0x80))); // 50% transparencency
   _color_scheme.insert(ColorScheme::value_type(RenderColor::AXES_COLOR, Color4f(0x00, 0x00, 0x00)));
   _color_scheme.insert(ColorScheme::value_type(RenderColor::OPENCSG_FACE_FRONT_COLOR, Color4f(0xf9, 0xd7, 0x2c)));
   _color_scheme.insert(ColorScheme::value_type(RenderColor::OPENCSG_FACE_BACK_COLOR, Color4f(0x9d, 0xcb, 0x51)));
@@ -58,6 +59,7 @@ RenderColorScheme::RenderColorScheme(fs::path path) : _path(path)
 	_show_in_gui = pt.get<bool>("show-in-gui");
 	
 	addColor(RenderColor::BACKGROUND_COLOR, "background");
+	addColor(RenderColor::HIGHLIGHT_COLOR, "highlight");
 	addColor(RenderColor::AXES_COLOR, "axes-color");
 	addColor(RenderColor::OPENCSG_FACE_FRONT_COLOR, "opencsg-face-front");
 	addColor(RenderColor::OPENCSG_FACE_BACK_COLOR, "opencsg-face-back");
@@ -132,6 +134,14 @@ void RenderColorScheme::addColor(RenderColor colorKey, std::string key)
     int g = (val >> 8) & 0xff;
     int b = val & 0xff;
     _color_scheme.insert(ColorScheme::value_type(colorKey, Color4f(r, g, b)));
+  } else if ((color.length() == 9) && (color.at(0) == '#')) {
+    char *endptr;
+    unsigned int val = strtol(color.substr(1).c_str(), &endptr, 16);
+    int r = (val >> 24) & 0xff;
+    int g = (val >> 16) & 0xff;
+    int b = (val >> 8) & 0xff;
+    int a = val & 0xff;
+    _color_scheme.insert(ColorScheme::value_type(colorKey, Color4f(r, g, b, a)));
   }
   else {
     throw std::invalid_argument(std::string("invalid color value for key '") + key + "': '" + color + "'");
diff --git a/src/colormap.h b/src/colormap.h
index a0dd9c350..2567f23d6 100644
--- a/src/colormap.h
+++ b/src/colormap.h
@@ -13,6 +13,7 @@ namespace fs = boost::filesystem;
 
 enum class RenderColor {
 	BACKGROUND_COLOR,
+	HIGHLIGHT_COLOR,
 	AXES_COLOR,
 	OPENCSG_FACE_FRONT_COLOR,
 	OPENCSG_FACE_BACK_COLOR,
diff --git a/src/mainwin.cc b/src/mainwin.cc
index c5edddeec..6819854d7 100755
--- a/src/mainwin.cc
+++ b/src/mainwin.cc
@@ -445,6 +445,7 @@ MainWindow::MainWindow(const QStringList &filenames)
 	connect(this->viewActionViewAll, SIGNAL(triggered()), this, SLOT(viewAll()));
 	connect(this->viewActionPerspective, SIGNAL(triggered()), this, SLOT(viewPerspective()));
 	connect(this->viewActionOrthogonal, SIGNAL(triggered()), this, SLOT(viewOrthogonal()));
+	connect(this->viewActionAnaglyph, SIGNAL(triggered()), this, SLOT(viewAnaglyph()));
 	connect(this->viewActionZoomIn, SIGNAL(triggered()), qglview, SLOT(ZoomIn()));
 	connect(this->viewActionZoomOut, SIGNAL(triggered()), qglview, SLOT(ZoomOut()));
 	connect(this->viewActionHideEditorToolBar, SIGNAL(triggered()), this, SLOT(hideEditorToolbar()));
@@ -752,10 +753,12 @@ void MainWindow::loadViewSettings(){
 		viewModeShowCrosshairs();
 	}
 	if (settings.value("view/showScaleProportional", true).toBool()) {
-        viewActionShowScaleProportional->setChecked(true);
-        viewModeShowScaleProportional();
-    }
-	if (settings.value("view/orthogonalProjection").toBool()) {
+		viewActionShowScaleProportional->setChecked(true);
+		viewModeShowScaleProportional();
+	}
+	if (settings.value("view/anaglyphProjection").toBool()) {
+		viewAnaglyph();
+	} else if (settings.value("view/orthogonalProjection").toBool()) {
 		viewOrthogonal();
 	} else {
 		viewPerspective();
@@ -2773,9 +2776,13 @@ void MainWindow::viewPerspective()
 {
 	QSettingsCached settings;
 	settings.setValue("view/orthogonalProjection",false);
+	settings.setValue("view/anaglyphProjection",false);
 	viewActionPerspective->setChecked(true);
 	viewActionOrthogonal->setChecked(false);
+	viewActionAnaglyph->setChecked(false);
+	this->top_ctx->set_variable("$anaglyph", Value(false));
 	this->qglview->setOrthoMode(false);
+	this->qglview->setAnaglyphMode(false);
 	this->qglview->updateGL();
 }
 
@@ -2783,9 +2790,28 @@ void MainWindow::viewOrthogonal()
 {
 	QSettingsCached settings;
 	settings.setValue("view/orthogonalProjection",true);
+	settings.setValue("view/anaglyphProjection",false);
+	viewActionPerspective->setChecked(true);
 	viewActionPerspective->setChecked(false);
 	viewActionOrthogonal->setChecked(true);
+	viewActionAnaglyph->setChecked(false);
+	this->top_ctx->set_variable("$anaglyph", Value(false));
 	this->qglview->setOrthoMode(true);
+	this->qglview->setAnaglyphMode(false);
+	this->qglview->updateGL();
+}
+
+void MainWindow::viewAnaglyph()
+{
+	QSettingsCached settings;
+	settings.setValue("view/orthogonalProjection",false);
+	settings.setValue("view/anaglyphProjection",true);
+	viewActionPerspective->setChecked(false);
+	viewActionOrthogonal->setChecked(false);
+	viewActionAnaglyph->setChecked(true);
+	this->top_ctx->set_variable("$anaglyph", Value(true));
+	this->qglview->setOrthoMode(false);
+	this->qglview->setAnaglyphMode(true);
 	this->qglview->updateGL();
 }
 
diff --git a/src/renderer.cc b/src/renderer.cc
index 7964e1961..b3b9e12ad 100644
--- a/src/renderer.cc
+++ b/src/renderer.cc
@@ -241,6 +241,7 @@ void Renderer::setColorScheme(const ColorScheme &cs) {
 	colormap[ColorMode::MATERIAL_EDGES] = ColorMap::getColor(cs, RenderColor::CGAL_EDGE_FRONT_COLOR);
 	colormap[ColorMode::CUTOUT_EDGES] = ColorMap::getColor(cs, RenderColor::CGAL_EDGE_BACK_COLOR);
 	colormap[ColorMode::EMPTY_SPACE] = ColorMap::getColor(cs, RenderColor::BACKGROUND_COLOR);
+	colormap[ColorMode::HIGHLIGHT] = ColorMap::getColor(cs, RenderColor::HIGHLIGHT_COLOR);
 	this->colorscheme = &cs;
 }
 
