diff --git a/color-schemes/render/ash.json b/color-schemes/render/ash.json
new file mode 100644
index 000000000..5b8fc8257
--- /dev/null
+++ b/color-schemes/render/ash.json
@@ -0,0 +1,20 @@
+{
+    "name" : "Ash",
+    "index" : 2000,
+    "show-in-gui" : true,
+    "description" : "monochrome version of Cornfield",
+
+    "colors" : {
+        "background" :         "#fdfdfd",
+        "axes-color" :         "#000000",
+        "opencsg-face-front" : "#d1d1d1",
+        "opencsg-face-back" :  "#b8b8b8",
+        "cgal-face-front" :    "#d1d1d1",
+        "cgal-face-back" :     "#b8b8b8",
+        "cgal-face-2d" :       "#939393",
+        "cgal-edge-front" :    "#e5e5e5",
+        "cgal-edge-back" :     "#c5c5c5",
+        "cgal-edge-2d" :       "#363636",
+        "crosshair" :          "#1b1b1b"
+    }
+}
diff --git a/images/Anaglyph-32.png b/images/Anaglyph-32.png
new file mode 100644
index 000000000..b7bc5c4a0
Binary files /dev/null and b/images/Anaglyph-32.png differ
diff --git a/openscad.qrc b/openscad.qrc
index 0b143032c..247305a3b 100644
--- a/openscad.qrc
+++ b/openscad.qrc
@@ -106,6 +106,7 @@
     <file>images/SVG-white.png</file>
     <file>images/PDF.png</file>
     <file>images/PDF-white.png</file>
+    <file>images/Anaglyph-32.png</file>
     <file>sounds/complete.wav</file>
   </qresource>
 </RCC>
diff --git a/src/Camera.cc b/src/Camera.cc
index 4c13820d6..e9882d28c 100644
--- a/src/Camera.cc
+++ b/src/Camera.cc
@@ -1,22 +1,29 @@
 #include "Camera.h"
+#include "QSettingsCached.h"
 #include "rendersettings.h"
 #include "printutils.h"
 #include "degree_trig.h"
 
 static const double DEFAULT_DISTANCE = 140.0;
 static const double DEFAULT_FOV = 22.5;
+static const float DEFAULT_EYE_SEPARATION = 15.0; // 65 mm avg. between human eyes
 
 Camera::Camera() :
 	projection(ProjectionType::PERSPECTIVE), fov(DEFAULT_FOV), viewall(false), autocenter(false)
 {
 	PRINTD("Camera()");
 
+	QSettingsCached settings;
 	// gimbal cam values
 	resetView();
 
 	pixel_width = RenderSettings::inst()->img_width;
 	pixel_height = RenderSettings::inst()->img_height;
 	locked = false;
+	anaglyph = false;
+	eyeSeparation = settings.value("3dview/eyeSeparation", DEFAULT_EYE_SEPARATION).toDouble();
+	if ((eyeSeparation <= 0.0) || (eyeSeparation > 100.0)) eyeSeparation = DEFAULT_EYE_SEPARATION;
+	std::cerr << "eye separation: " << this->eyeSeparation << std::endl;
 }
 
 void Camera::setup(std::vector<double> params)
@@ -65,6 +72,16 @@ void Camera::viewAll(const BoundingBox &bbox)
 	}
 }
 
+void Camera::zoomAnaglyph(int zoom)
+{
+	QSettingsCached settings;
+	this->eyeSeparation *= pow(0.95, -zoom / 120.0);
+	if (eyeSeparation >= 100.0) eyeSeparation = 100.0;
+	else if (eyeSeparation <= 0.0) eyeSeparation = DEFAULT_EYE_SEPARATION;
+	settings.setValue("3dview/eyeSeparation", QString::fromStdString(std::to_string(this->eyeSeparation)));
+	std::cerr << "eye separation: " << this->eyeSeparation << std::endl;
+}
+
 void Camera::zoom(int zoom, bool relative)
 {
     if (relative) {
diff --git a/src/Camera.h b/src/Camera.h
index 4cd2859c3..b770f6d24 100644
--- a/src/Camera.h
+++ b/src/Camera.h
@@ -29,6 +29,7 @@ public:
 	void setup(std::vector<double> params);
 	void gimbalDefaultTranslate();
 	void setProjection(ProjectionType type);
+	void zoomAnaglyph(int delta);
 	void zoom(int delta, bool relative);
 	double zoomValue() const;
 	double fovValue() const;
@@ -67,6 +68,10 @@ public:
 	// (--camera option in commandline mode)
 	bool locked;
 
+	// true if in anaglyph 3D stereo view
+	bool anaglyph;
+	double eyeSeparation;
+
 protected:
 	// Perspective settings
 	double viewer_distance;
diff --git a/src/GLView.cc b/src/GLView.cc
index 766387736..43631fffb 100644
--- a/src/GLView.cc
+++ b/src/GLView.cc
@@ -115,8 +115,104 @@ void GLView::setupCamera() const
   glRotated(cam.object_rot.z(), 0.0, 0.0, 1.0);
 }
 
+/* set up camera for left or right eye */
+/* after https://quiescentspark.blogspot.com/2011/05/rendering-3d-anaglyph-in-opengl.html */
+void GLView::setup3dCamera(bool leftCamera) {
+  float left, right, bottom, top;
+  float convergence = (float)cam.zoomValue();
+  float aspectRatio = aspectratio;
+  float fov = cam.fov;
+  float nearClippingDistance = 0.1 * cam.zoomValue();
+  float farClippingDistance = 100.0 * cam.zoomValue();
+
+  // calculate frustum
+  top = nearClippingDistance * tan_degrees(fov / 2.0);
+  bottom = -top;
+  float a = aspectRatio * tan_degrees(fov / 2.0) * convergence;
+  float b = a - cam.eyeSeparation / 2;
+  float c = a + cam.eyeSeparation / 2;
+  if (leftCamera) { // left camera
+    left = -b * nearClippingDistance / convergence;
+    right = c * nearClippingDistance / convergence;
+  }
+  else { // right camera
+    left = -c * nearClippingDistance / convergence;
+    right = b * nearClippingDistance / convergence;
+  }
+
+  glMatrixMode(GL_PROJECTION);
+  glLoadIdentity();
+  glFrustum(left, right, bottom, top, nearClippingDistance, farClippingDistance);
+  glMatrixMode(GL_MODELVIEW);
+  glLoadIdentity();
+  if (leftCamera) { // left camera
+    glTranslatef(cam.eyeSeparation / 2, 0.0, 0.0);
+  }
+  else { // right camera
+    glTranslatef(-cam.eyeSeparation / 2, 0.0, 0.0);
+  }
+  gluLookAt(0.0, -cam.zoomValue(), 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0);
+  glRotated(cam.object_rot.x(), 1.0, 0.0, 0.0);
+  glRotated(cam.object_rot.y(), 0.0, 1.0, 0.0);
+  glRotated(cam.object_rot.z(), 0.0, 0.0, 1.0);
+}
+
+void GLView::paintOnce()
+{
+  auto crosshaircol = ColorMap::getColor(*this->colorscheme, RenderColor::CROSSHAIR_COLOR);
+  // The crosshair should be fixed at the center of the viewport...
+  if (showcrosshairs) GLView::showCrosshairs(crosshaircol);
+  glTranslated(cam.object_trans.x(), cam.object_trans.y(), cam.object_trans.z());
+  // ...the axis lines need to follow the object translation.
+  auto axescolor = ColorMap::getColor(*this->colorscheme, RenderColor::AXES_COLOR);
+  if (showaxes) GLView::showAxes(axescolor);
+  // mark the scale along the axis lines
+  if (showaxes && showscale) GLView::showScalemarkers(axescolor);
+
+  glEnable(GL_LIGHTING);
+  glDepthFunc(GL_LESS);
+  glCullFace(GL_BACK);
+  glDisable(GL_CULL_FACE);
+  glLineWidth(2);
+  glColor3d(1.0, 0.0, 0.0);
+
+  if (this->renderer) {
+#if defined(ENABLE_OPENCSG)
+    // FIXME: This belongs in the OpenCSG renderer, but it doesn't know about this ID yet
+    OpenCSG::setContext(this->opencsg_id);
+#endif
+    this->renderer->prepare(showfaces, showedges);
+    this->renderer->draw(showfaces, showedges);
+  }
+
+  glDisable(GL_LIGHTING);
+}
+
+void GLView::anaglyphHalfColor() {
+  glDisable(GL_LIGHTING);
+  auto bgcol = ColorMap::getColor(*this->colorscheme, RenderColor::BACKGROUND_COLOR);
+  glClearColor(bgcol[0], bgcol[1], bgcol[2], 1.0);
+
+  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+  setup3dCamera(true);
+  glColorMask(true, false, false, false); // red
+  paintOnce();
+
+  glClear(GL_DEPTH_BUFFER_BIT);
+  setup3dCamera(false);
+  glColorMask(false, true, true, false); // cyan
+  paintOnce();
+
+  glColorMask(true, true, true, true);
+}
+
 void GLView::paintGL()
 {
+  if (cam.anaglyph) {
+    anaglyphHalfColor();
+    return;
+  }
+
   glDisable(GL_LIGHTING);
   auto bgcol = ColorMap::getColor(*this->colorscheme, RenderColor::BACKGROUND_COLOR);
   auto axescolor = ColorMap::getColor(*this->colorscheme, RenderColor::AXES_COLOR);
diff --git a/src/GLView.h b/src/GLView.h
index 1fcd31bc8..4df9560bf 100644
--- a/src/GLView.h
+++ b/src/GLView.h
@@ -2,20 +2,20 @@
 
 /* GLView: A basic OpenGL rectangle for rendering images.
 
-This class is inherited by:
+	 This class is inherited by:
 
-*QGLview - for Qt GUI
-*OffscreenView - for offscreen rendering, in tests and from command-line
-(This class is also overridden by NULLGL.cc for special experiments)
+ * QGLview - for Qt GUI
+ * OffscreenView - for offscreen rendering, in tests and from command-line
+	 (This class is also overridden by NULLGL.cc for special experiments)
 
-The view assumes either a Gimbal Camera (rotation,translation,distance)
-or Vector Camera (eye,center/target) is being used. See Camera.h. The
-cameras are not kept in sync.
+	 The view assumes either a Gimbal Camera (rotation,translation,distance)
+	 or Vector Camera (eye,center/target) is being used. See Camera.h. The
+	 cameras are not kept in sync.
 
-QGLView only uses GimbalCamera while OffscreenView can use either one.
-Some actions (showCrossHairs) only work properly on Gimbal Camera.
+	 QGLView only uses GimbalCamera while OffscreenView can use either one.
+	 Some actions (showCrossHairs) only work properly on Gimbal Camera.
 
-*/
+ */
 
 #include <Eigen/Core>
 #include <Eigen/Geometry>
@@ -29,7 +29,7 @@ class GLView
 {
 public:
 	GLView();
-	void setRenderer(class Renderer* r);
+	void setRenderer(class Renderer *r);
 	Renderer *getRenderer() const { return this->renderer; }
 
 	void initializeGL();
@@ -39,6 +39,10 @@ public:
 	void setCamera(const Camera &cam);
 	void setupCamera() const;
 
+  void setup3dCamera(bool leftCamera);
+  void paintOnce();
+  void anaglyphHalfColor();
+
 	void setColorScheme(const ColorScheme &cs);
 	void setColorScheme(const std::string &cs);
 	void updateColorScheme();
@@ -58,7 +62,7 @@ public:
 	virtual std::string getRendererInfo() const = 0;
 	virtual float getDPI() { return 1.0f; }
 
-	virtual ~GLView(){};
+	virtual ~GLView() {}
 
 	Renderer *renderer;
 	const ColorScheme *colorscheme;
diff --git a/src/MainWindow.h b/src/MainWindow.h
index aaed0e139..25f08b381 100644
--- a/src/MainWindow.h
+++ b/src/MainWindow.h
@@ -299,6 +299,7 @@ public slots:
 	void viewCenter();
 	void viewPerspective();
 	void viewOrthogonal();
+	void viewAnaglyph();
 	void viewTogglePerspective();
 	void viewResetView();
 	void viewAll();
diff --git a/src/MainWindow.ui b/src/MainWindow.ui
index c3ea8e0e9..93eede35f 100644
--- a/src/MainWindow.ui
+++ b/src/MainWindow.ui
@@ -251,6 +251,7 @@
          <addaction name="separator"/>
          <addaction name="viewActionPerspective"/>
          <addaction name="viewActionOrthogonal"/>
+         <addaction name="viewActionAnaglyph"/>
          <addaction name="separator"/>
          <addaction name="viewActionShowAxes"/>
          <addaction name="viewActionShowScaleProportional"/>
@@ -413,6 +414,7 @@
     <addaction name="separator"/>
     <addaction name="viewActionPerspective"/>
     <addaction name="viewActionOrthogonal"/>
+    <addaction name="viewActionAnaglyph"/>
     <addaction name="separator"/>
     <addaction name="viewActionHideEditorToolBar"/>
     <addaction name="viewActionHide3DViewToolBar"/>
@@ -1302,6 +1304,24 @@
     <string>&amp;Orthogonal</string>
    </property>
   </action>
+  <action name="viewActionAnaglyph">
+   <property name="checkable">
+    <bool>true</bool>
+   </property>
+   <property name="icon">
+    <iconset resource="../openscad.qrc">
+     <normaloff>:/images/Anaglyph-32.png</normaloff>:/images/Anaglyph-32.png</iconset>
+   </property>
+   <property name="text">
+    <string>&amp;Stereo Vision (Anaglyph)</string>
+   </property>
+   <property name="toolTip">
+    <string>Stereo Vision (Anaglyph) - requires special red-cyan glasses</string>
+   </property>
+   <property name="shortcut">
+    <string>Alt+S</string>
+   </property>
+  </action>
   <action name="windowActionHideConsole">
    <property name="checkable">
     <bool>true</bool>
diff --git a/src/QGLView.cc b/src/QGLView.cc
index bce613073..c2d7f0b6c 100644
--- a/src/QGLView.cc
+++ b/src/QGLView.cc
@@ -324,7 +324,10 @@ void QGLView::wheelEvent(QWheelEvent *event)
 {
 	const auto pos = event->pos();
 	const int v = event->angleDelta().y();
-	if (this->mouseCentricZoom) {
+	if (QGuiApplication::keyboardModifiers().testFlag(Qt::ControlModifier)) {
+		cam.zoomAnaglyph(v);
+		updateGL();
+	} else if (this->mouseCentricZoom) {
 		zoomCursor(pos.x(), pos.y(), v);
 	} else {
 		zoom(v, true);
diff --git a/src/QGLView.h b/src/QGLView.h
index 195852972..33325b762 100644
--- a/src/QGLView.h
+++ b/src/QGLView.h
@@ -41,6 +41,7 @@ public:
 	void setOrthoMode(bool enabled);
 	bool showScaleProportional() const { return this->showscale; }
 	void setShowScaleProportional(bool enabled) { this->showscale = enabled; }
+	void setAnaglyphMode(bool enabled) { this->cam.anaglyph = enabled; }
 	std::string getRendererInfo() const override;
 	float getDPI() override { return this->devicePixelRatio(); }
 
diff --git a/src/builtin.cc b/src/builtin.cc
index aa857908a..6f20d6b54 100644
--- a/src/builtin.cc
+++ b/src/builtin.cc
@@ -113,6 +113,7 @@ Builtins::Builtins()
 	this->assignments.emplace_back(new Assignment("$vpr", make_shared<Literal>(VectorType(0.0, 0.0, 0.0))) );
 	this->assignments.emplace_back(new Assignment("$vpd", make_shared<Literal>(500)) );
 	this->assignments.emplace_back(new Assignment("$vpf", make_shared<Literal>(22.5)) );
+	this->assignments.emplace_back(new Assignment("$anaglyph", make_shared<Literal>(Value::undefined.clone())) ); // true if rendering anaglyph 3d
 }
 
 void Builtins::initKeywordList()
diff --git a/src/mainwin.cc b/src/mainwin.cc
index c5edddeec..6819854d7 100755
--- a/src/mainwin.cc
+++ b/src/mainwin.cc
@@ -445,6 +445,7 @@ MainWindow::MainWindow(const QStringList &filenames)
 	connect(this->viewActionViewAll, SIGNAL(triggered()), this, SLOT(viewAll()));
 	connect(this->viewActionPerspective, SIGNAL(triggered()), this, SLOT(viewPerspective()));
 	connect(this->viewActionOrthogonal, SIGNAL(triggered()), this, SLOT(viewOrthogonal()));
+	connect(this->viewActionAnaglyph, SIGNAL(triggered()), this, SLOT(viewAnaglyph()));
 	connect(this->viewActionZoomIn, SIGNAL(triggered()), qglview, SLOT(ZoomIn()));
 	connect(this->viewActionZoomOut, SIGNAL(triggered()), qglview, SLOT(ZoomOut()));
 	connect(this->viewActionHideEditorToolBar, SIGNAL(triggered()), this, SLOT(hideEditorToolbar()));
@@ -752,10 +753,12 @@ void MainWindow::loadViewSettings(){
 		viewModeShowCrosshairs();
 	}
 	if (settings.value("view/showScaleProportional", true).toBool()) {
-        viewActionShowScaleProportional->setChecked(true);
-        viewModeShowScaleProportional();
-    }
-	if (settings.value("view/orthogonalProjection").toBool()) {
+		viewActionShowScaleProportional->setChecked(true);
+		viewModeShowScaleProportional();
+	}
+	if (settings.value("view/anaglyphProjection").toBool()) {
+		viewAnaglyph();
+	} else if (settings.value("view/orthogonalProjection").toBool()) {
 		viewOrthogonal();
 	} else {
 		viewPerspective();
@@ -2773,9 +2776,13 @@ void MainWindow::viewPerspective()
 {
 	QSettingsCached settings;
 	settings.setValue("view/orthogonalProjection",false);
+	settings.setValue("view/anaglyphProjection",false);
 	viewActionPerspective->setChecked(true);
 	viewActionOrthogonal->setChecked(false);
+	viewActionAnaglyph->setChecked(false);
+	this->top_ctx->set_variable("$anaglyph", Value(false));
 	this->qglview->setOrthoMode(false);
+	this->qglview->setAnaglyphMode(false);
 	this->qglview->updateGL();
 }
 
@@ -2783,9 +2790,28 @@ void MainWindow::viewOrthogonal()
 {
 	QSettingsCached settings;
 	settings.setValue("view/orthogonalProjection",true);
+	settings.setValue("view/anaglyphProjection",false);
+	viewActionPerspective->setChecked(true);
 	viewActionPerspective->setChecked(false);
 	viewActionOrthogonal->setChecked(true);
+	viewActionAnaglyph->setChecked(false);
+	this->top_ctx->set_variable("$anaglyph", Value(false));
 	this->qglview->setOrthoMode(true);
+	this->qglview->setAnaglyphMode(false);
+	this->qglview->updateGL();
+}
+
+void MainWindow::viewAnaglyph()
+{
+	QSettingsCached settings;
+	settings.setValue("view/orthogonalProjection",false);
+	settings.setValue("view/anaglyphProjection",true);
+	viewActionPerspective->setChecked(false);
+	viewActionOrthogonal->setChecked(false);
+	viewActionAnaglyph->setChecked(true);
+	this->top_ctx->set_variable("$anaglyph", Value(true));
+	this->qglview->setOrthoMode(false);
+	this->qglview->setAnaglyphMode(true);
 	this->qglview->updateGL();
 }
 
